bee-storage/bee-storage-rocksdb/src/access/stream.rs

```diff
@@ -0,0 +1,350 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    column_families::*,
+    storage::{Storage, StorageBackend},
+    system::System,
+};
+
+use bee_common::packable::Packable;
+use bee_ledger::{
+    snapshot::info::SnapshotInfo,
+    types::{Balance, ConsumedOutput, CreatedOutput, LedgerIndex, OutputDiff, Receipt, TreasuryOutput, Unspent},
+};
+use bee_message::{
+    address::{Address, Ed25519Address},
+    milestone::{Milestone, MilestoneIndex},
+    output::OutputId,
+    payload::indexation::{PaddedIndex, INDEXATION_PADDED_INDEX_LENGTH},
+    Message, MessageId, MESSAGE_ID_LENGTH,
+};
+use bee_storage::access::AsStream;
+use bee_tangle::{
+    metadata::MessageMetadata, solid_entry_point::SolidEntryPoint, unconfirmed_message::UnconfirmedMessage,
+};
+
+use futures::{
+    stream::Stream,
+    task::{Context, Poll},
+};
+use pin_project::pin_project;
+use rocksdb::{DBIterator, IteratorMode};
+
+use std::{convert::TryInto, marker::PhantomData, pin::Pin};
+
+#[pin_project(project = StorageStreamProj)]
+pub struct StorageStream<'a, K, V> {
+    #[pin]
+    inner: DBIterator<'a>,
+    budget: usize,
+    counter: usize,
+    marker: PhantomData<(K, V)>,
+}
+
+impl<'a, K, V> StorageStream<'a, K, V> {
+    fn new(inner: DBIterator<'a>, budget: usize) -> Self {
+        StorageStream::<K, V> {
+            inner,
+            budget,
+            counter: 0,
+            marker: PhantomData,
+        }
+    }
+}
+
+macro_rules! impl_stream {
+    ($key:ty, $value:ty, $cf:expr) => {
+        #[async_trait::async_trait]
+        impl<'a> AsStream<'a, $key, $value> for Storage {
+            type Stream = StorageStream<'a, $key, $value>;
+
+            async fn stream(&'a self) -> Result<Self::Stream, <Self as StorageBackend>::Error> {
+                Ok(StorageStream::new(
+                    self.inner.iterator_cf(self.cf_handle($cf)?, IteratorMode::Start),
+                    self.config.iteration_budget,
+                ))
+            }
+        }
+
+        impl<'a> Stream for StorageStream<'a, $key, $value> {
```

Comment:
I'd add a tiny docstring explaining what this stream is supposed to iterate over